server/src/main/java/io/unitycatalog/server/service/AuthService.java

```diff
@@ -0,0 +1,158 @@
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.ToString;
+import lombok.extern.jackson.Jacksonized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ExceptionHandler(GlobalExceptionHandler.class)
+public class AuthService {
+
+  interface Fields {
+    String GRANT_TYPE = "grant_type";
+    String CLIENT_ID = "client_id";
+    String CLIENT_SECRET = "client_secret";
+    String SUBJECT_TOKEN = "subject_token";
+    String SUBJECT_TOKEN_TYPE = "subject_token_type";
+    String ACTOR_TOKEN = "actor_token";
+    String ACTOR_TOKEN_TYPE = "actor_token_type";
+    String REQUESTED_TOKEN_TYPE = "requested_token_type";
+  }
+
+  interface GrantTypes {
+    String TOKEN_EXCHANGE = "urn:ietf:params:oauth:grant-type:token-exchange";
+  }
+
+  interface TokenTypes {
+    String ACCESS = "urn:ietf:params:oauth:token-type:access_token";
+    String ID = "urn:ietf:params:oauth:token-type:id_token";
+    String JWT = "urn:ietf:params:oauth:token-type:jwt";
+  }
+
+  interface AuthTypes {
+    String BEARER = "Bearer";
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AuthService.class);
+
+  private final SecurityContext securityContext;
+  private final JwksOperations jwksOperations;
+
+  public AuthService(SecurityContext securityContext) {
+    this.securityContext = securityContext;
+    this.jwksOperations = new JwksOperations();
+  }
+
+  /**
+   * OAuth token exchange.
+   *
+   * <p>Performs an OAuth token exchange for an access-token.
+   *
+   * <p>TODO: This could probably be integrated into the OpenAPI spec.
+   *
+   * @param request The token exchange parameters
+   * @return The token exchange response
+   */
+  @Post("/tokens")
+  public HttpResponse grantToken(OAuthTokenExchangeRequest request) {
+    LOGGER.debug("Got token: {}", request);
+
+    if (request.getGrantType() != null
+        && !TokenTypes.ACCESS.equals(request.getRequestedTokenType())) {
+      throw new OAuthInvalidRequestException(
+          ErrorCode.INVALID_ARGUMENT,
+          "Unsupported requested token type: " + request.getRequestedTokenType());
+    }
+
+    if (request.getSubjectTokenType() == null) {
+      throw new OAuthInvalidRequestException(
+          ErrorCode.INVALID_ARGUMENT, "Subject token type is required but was not specified");
+    }
+
+    if (request.getActorTokenType() != null) {
+      throw new OAuthInvalidRequestException(
+          ErrorCode.INVALID_ARGUMENT, "Actor tokens not currently supported");
+    }
+
+    DecodedJWT decodedJWT = JWT.decode(request.getSubjectToken());
+    String issuer = decodedJWT.getClaim("iss").asString();
+    String keyId = decodedJWT.getHeaderClaim("kid").asString();
+
+    LOGGER.debug("Validating token for issuer: {}", issuer);
+
+    JWTVerifier jwtVerifier = jwksOperations.verifierForIssuerAndKey(issuer, keyId);
+    decodedJWT = jwtVerifier.verify(decodedJWT);
+
+    String accessToken = securityContext.createAccessToken(decodedJWT);
```

Comment:
Should this access token have a limited lifetime for security?

Reply:
We are planning on addressing this in the next release along with token refresh. We agreed that for the short term, it is sufficient to invalidate existing tokens by rotating signing keys (by deleting and recreating them)

Reply:
Should we create an issue for this?